fixes #6792, use a simpler composite in condition implementation

nineinchnick opened the pull request:
The `(a,b) IN ((1,2))` notation is not supported by MySQL/MariaDB and SQLite, only PostgreSQL.

fixes #6792

nineinchnick committed:
fixes #6792, use a simpler composite in condition implementation on dbs other than pgsql

qiangxue commented:
> The (a,b) IN ((1,2)) notation is not supported by MySQL/MariaDB

It works well for MySQL 5.1 and above. See http://sqlfiddle.com/#!2/72132/4/0
It doesn't work with SQLite, though.

nineinchnick commented:
Right, somehow I didn't find it in the docs before and didn't actually run the query. Should we test other databases before deciding which implementation should be default? 

For reference, it's the `ROW()` expression syntax mentioned on the [expressions manual page](http://dev.mysql.com/doc/refman/5.1/en/expressions.html).

Now it occured to me that maybe a more verbose `ROW()` form should be used instead of `()` only, like `ROW(a, b) IN (ROW(1,2)). Should I make a separate PR for this?

qiangxue commented:
Yes, we should test other DBMS (sqlfiddle should allow us to try different DBMS, but I didn't have time to try them).

nineinchnick committed:
override only mssql and sqlite QueryBuilder; throw an exception for composite in condition with a subquery

nineinchnick commented:
After some research, I've updated the PR and here are my findings.

Only MSSQL and SQLite doesn't support this notation.

References:

* for MSSQL: the manual for [in condition](https://msdn.microsoft.com/en-us/library/ms177682.aspx) and [expressions](https://msdn.microsoft.com/en-us/library/ms190286.aspx) and [an article about in condition](http://explainextended.com/2009/06/17/efficient-exists/).
* fro SQLite: [the manual for expressions](http://sqlite.org/lang_expr.html)

The composite `IN` condition for a list of values can be implemented using simple `OR/AND` operators. There is a problem with subqueries. It is possible, but require aliased column names to be used in the condition. Since I don't think it's possible to extract a valid alias to prepend to column names I added throwing exceptions in that case.

The solution would be to build a `EXISTS (SELECT 1 FROM (subquery) a WHERE a.column1 = topQuery.column1 AND a.column2 = topQuery.column2)` condition, but there's a problem with the `topQuery` alias. The `EXISTS (subquery)` is supported by both MSSQL and SQLite.

nineinchnick commented on framework/db/QueryBuilder.php:
```diff
@@ -1067,22 +1067,7 @@ public function buildInCondition($operator, $operands, &$params)
         if ($values instanceof Query) {
-            // sub-query
-            list($sql, $params) = $this->build($values, $params);
-            $column = (array) $column;
```
This was a bug, making the `else` part below never to execute. I've removed this cast.

nineinchnick commented on tests/unit/framework/db/QueryBuilderTest.php:
```diff
@@ -228,6 +222,26 @@ public function conditionProvider()
+                    [ ['in', ['id', 'name'], [['id' => 1, 'name' => 'foo'], ['id' => 2, 'name' => 'bar']]], '((`id` = :qp0 AND `name` = :qp1) OR (`id` = :qp2 AND `name` = :qp3))', [':qp0' => 1, ':qp1' => 'foo', ':qp2' => 2, ':qp3' => 'bar'] ],
+                    [ ['not in', ['id', 'name'], [['id' => 1, 'name' => 'foo'], ['id' => 2, 'name' => 'bar']]], '((`id` != :qp0 OR `name` != :qp1) AND (`id` != :qp2 OR `name` != :qp3))', [':qp0' => 1, ':qp1' => 'foo', ':qp2' => 2, ':qp3' => 'bar'] ],
+                    //[ ['in', ['id', 'name'], (new Query())->select(['id', 'name'])->from('users')->where(['active' => 1])], 'EXISTS (SELECT 1 FROM (SELECT `id`, `name` FROM `users` WHERE `active`=:qp0) AS a WHERE a.`id` = `id AND a.`name` = `name`)', [':qp0' => 1] ],
+                    //[ ['not in', ['id', 'name'], (new Query())->select(['id', 'name'])->from('users')->where(['active' => 1])], 'NOT EXISTS (SELECT 1 FROM (SELECT `id`, `name` FROM `users` WHERE `active`=:qp0) AS a WHERE a.`id` = `id` AND a.`name = `name`)', [':qp0' => 1] ],
```
Should I remove those? We still could try to implement this.

qiangxue replied:
How can we implement it?

nineinchnick replied:
Creating a query like this would work:
~~~sql
EXISTS (
  SELECT 1
  FROM (
    SELECT c1, c2 FROM t1
  ) AS a
  WHERE a.c1 = top.c1 AND a.c2 = top.c2
)
~~~
There is a problem however how to extract the `top` alias name from the main query. We could try to parse the `from` property but there could be more than one table name there and there is no way to tell which one is the main one.
We could ignore that fact and just assume that columns passed in the condition should be aliased, but if they are not this fails silently, that is invalid results are produced and it will be hard to debug.

In the code above the `id` and `name` columns are not aliased and actually there isn't even a `from` value.

qiangxue replied:
That's a bit too complicated. Let's leave it for now until it is requested by many people.

nineinchnick commented on tests/unit/framework/db/QueryBuilderTest.php:
```diff
@@ -444,4 +458,17 @@ public function testSelectSubquery()
+                ['id' => 2, 'name' => 'bar'],
+            ],
+        ];
+        (new Query())->from('customer')->where($condition)->all($this->getConnection());
```
I had to add this test to make it fail before adding a working implementation.

qiangxue commented:
Good job!

qiangxue commented:
Thank you for your excellent work!
